src/refract/PrintVisitor.cc

```diff
@@ -6,180 +6,179 @@
 
-                refract::Visit(*this, *(*i));
-            }
-
-            indentOS(indent);
-            os << "}\n";
+        for (const auto& m : e.meta) {
+            PrintVisitor{indent + 1, os, ommitSrcMap}(*m);
         }
     }
 
     void PrintVisitor::printAttr(const IElement& e)
     {
-        if (e.attributes.size() > 0) {
-            os << "attributes {\n";
-            indentOS(indent);
-
-            for (IElement::MemberElementCollection::const_iterator i = e.attributes.begin();
-                i != e.attributes.end();
-                ++i) {
+        indented() << "- <attr>\n";
 
-                refract::Visit(*this, *(*i));
-            }
-
-            indentOS(indent);
-            os << "}\n";
+        for (const auto& a : e.attributes) {
+            PrintVisitor{indent + 1, os, ommitSrcMap}(*a);
         }
     }
 
-
     void PrintVisitor::operator()(const IElement& e)
     {
-        indentOS(indent);
-        printMeta(e);
-        printAttr(e);
+        indented() << "+ " << e.element() << '\n';
+
+        PrintVisitor pv{indent + 1, os, ommitSrcMap};
+
+        refract::VisitBy(e, pv);
 
-        VisitBy(e, *this);
+        pv.printMeta(e);
+        pv.printAttr(e);
     }
 
     void PrintVisitor::operator()(const NullElement& e)
     {
-        os << "NullElement(null)" << "\n";
+        indented() << "- Null\n";
     }
 
     void PrintVisitor::operator()(const HolderElement& e)
     {
-        os << "Direct Element(" << e.element() << ") {\n";
-        indentOS(indent + 1);
+        indented() << "- Holder[" << e.element() << "]\n";
 
-        PrintVisitor ps(indent + 1, os);
-        Visitor visitor(ps);
-
-        visitor.visit(*e.value);
-
-        indentOS(indent);
-        os << "}\n";
+        assert(e.value);
+        PrintVisitor{indent + 1, os, ommitSrcMap}(*e.value);
     }
 
     void PrintVisitor::operator()(const StringElement& e)
     {
         const StringElement::ValueType* v = GetValue<StringElement>(e);
-        os << "StringElement(" << *v << ")" << "\n";
+
+        assert(v);
+        indented() << "- String \"" << *v << "\"\n";
     }
 
     void PrintVisitor::operator()(const NumberElement& e)
     {
         const NumberElement::ValueType* v = GetValue<NumberElement>(e);
 
-        os << "NumberElement(";
+        auto& out = indented();
+        out << "- Number ";
         if (v) {
-            os << *v;
-        } else {
-            os << "null";
+            out << *v;
         }
-        os << ")" << "\n";
+        out << "\n";
     }
 
     void PrintVisitor::operator()(const BooleanElement& e)
     {
         const BooleanElement::ValueType* v = GetValue<BooleanElement>(e);
 
-        os << "BooleanElement(";
+        auto& out = indented();
+        out << "- Boolean ";
         if (v) {
-            os << *v;
-        } else {
-            os << "null";
+            out << *v;
         }
-        os << ")" << "\n";
+        out << "\n";
     }
```

Comment:
And, on a similar note, is it valid programmatically to have the first value of Member be of other type than String? If no, I'd suggest to gut the part where we're throwing an exception and make it an assert as well.